### PR TITLE
Validation for data stream creation

### PR DESCRIPTION
**Pull request by danhermann**
Prevent name conflicts with existing indices or aliases.

Relates to #53100

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java [hunk 1]
```diff
@@ -161,15 +165,63 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
...
+                logger.info("adding data stream [{}]", request.name);
+                return ClusterState.builder(currentState).metaData(builder).build();
+            } else {
+                ValidationException ex = new ValidationException();
+                ex.addValidationErrors(validationErrors);
+                throw new IllegalArgumentException(ex);
+            }
+        }
+
+        private static List<String> validateDataStreamName(String name) {
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java]: I think `MetaDataCreateIndexService.validateIndexName()` or `MetaDataCreateIndexService.validateIndexOrAliasName()` could be reused?

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 2]
```diff
@@ -1315,20 +1315,36 @@ public MetaData build() {
...
 
+            DataStreamMetadata dsMetadata = (DataStreamMetadata) customs.get(DataStreamMetadata.TYPE);
+            if (dsMetadata != null) {
+                for (DataStream ds : dsMetadata.dataStreams().values()) {
+                    if (aliasAndIndexLookup.containsKey(ds.getName())) {
+                        throw new IllegalStateException("data stream [" + ds.getName() + "] conflicts with existing index or alias");
+                    }
+
+                    final String backingIndexPrefix = (ds.getName().startsWith(".") ? "" : ".") + ds.getName() + "-";
+                    for (String indexName : allIndices) {
```
[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: I think we don't have to iterate over all indices and can reuse `aliasAndIndexLookup` sorted map.
If we use: `aliasAndIndexLookup.subMap(...)` then this returns all aliases / indices that share the
same prefix are returned and we can check if it is empty and if not throw an error. 

This method is also used in the `WildcardExpressionResolver#suffixWildcard(...)` method.

[the remark this answers is not in the document]
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: maybe also move this logic into a private method?

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 3]
```diff
@@ -1371,6 +1372,25 @@ public MetaData build() {
...
+                    if (aliasAndIndexLookup.containsKey(ds.getName())) {
+                        throw new IllegalStateException("data stream [" + ds.getName() + "] conflicts with existing index or alias");
+                    }
+
+                    final String backingIndexPrefixFrom = (ds.getName().startsWith(".") ? "" : ".") + ds.getName() + "-";
+                    final String backingIndexPrefixTo = (ds.getName().startsWith(".") ? "" : ".") + ds.getName() + ".";
+                    SortedMap<?, ?> map = aliasAndIndexLookup.subMap(backingIndexPrefixFrom, backingIndexPrefixTo);
+                    if (map.size() != 0) {
+                        throw new IllegalStateException(
+                            "data stream [" + ds.getName() + "] could create backing indices that conflict with existing indices");
```
Comment (martijnvg) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: nit: a backing index could also conflict with an alias (or data stream in the future)

**Review by martijnvg — APPROVED** [review]
LGTM

**Comment by martijnvg** [comment]
@elasticmachine run elasticsearch-ci/1

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 4]
```diff
@@ -1371,6 +1372,23 @@ public MetaData build() {
...
+            if (dsMetadata != null) {
+                for (DataStream ds : dsMetadata.dataStreams().values()) {
+                    if (aliasAndIndexLookup.containsKey(ds.getName())) {
+                        throw new IllegalStateException("data stream [" + ds.getName() + "] conflicts with existing index or alias");
+                    }
+
+                    SortedMap<?, ?> map = aliasAndIndexLookup.subMap(ds.getName() + "-", ds.getName() + "."); // '.' is the char after '-'
+                    if (map.size() != 0) {
+                        throw new IllegalStateException("data stream [" + ds.getName() +
+                            "] could create backing indices that conflict with existing indices or aliases");
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Should we include the list of conflicting names in the message or at least the first few?

File: server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java [hunk 5]
```diff
@@ -165,9 +166,11 @@ static ClusterState createDataStream(ClusterState currentState, Request request)
                 throw new IllegalArgumentException("data_stream [" + request.name + "] already exists");
             }
 
+            MetaDataCreateIndexService.validateIndexOrAliasName(request.name,
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java]: Should we rename the method to either "validateIndexSpaceName" or "validateAbstractIndexName"? Can be done in a follow-up when the naming discussion is clarified.

File: server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java [hunk 6]
```diff
@@ -165,9 +166,11 @@ static ClusterState createDataStream(ClusterState currentState, Request request)
...
+                (s1, s2) -> new IllegalArgumentException("data_stream [" + s1 + "] " + s2));
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java]: The other usages of `validateIndexOrAliasName` throw a dedicated exception. I am not entirely sure if `IllegalArgumentException` results in the same RestStatus and message handling. Could we add this to the rest test suite?

**Review by henningandersen — APPROVED** [review]
LGTM.

Left a few minor comments to consider.

File: server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java (hunk 6 above)
Comment (danhermann) [reply server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java]: Added a REST test to enforce that the same HTTP 400 response code is returned.

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix-unix

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix-unix

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine update branch

**Pull request merged**